ocaml/typing/typemod.ml

```diff
@@ -2519,8 +2549,20 @@ and type_module_aux ~alias sttn funct_body anchor env smod =
       in
       md, shape
   | Pmod_structure sstr ->
+      let expected_sig = match expected_modtype with
+        | None -> None
+        | Some mty ->
+          (match mty with
+          | Mty_ident _
+          | Mty_functor _
+          | Mty_alias _
+          | Mty_strengthen _ -> None
+          (* For now, we only allow module type inference in cases where the signature
+             is "close" to the struct (i.e. declared in the same place). *)
```

Comment:
For things that we need to improve in the near future, let's add a CR comment block.